File: 0508/0508-285/sdk/go/config-client.go
package configclient

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"sync"
	"time"
)

type ConfigClient struct {
	serverURL   string
	appID       string
	namespace   string
	environment string
	tags        map[string]string
	configs     map[string]interface{}
	versions    map[string]int64
	mu          sync.RWMutex
	callbacks   []func(map[string]interface{})
}

func NewConfigClient(serverURL, appID, namespace, environment string) *ConfigClient {
	return &ConfigClient{
		serverURL:   serverURL,
		appID:       appID,
		namespace:   namespace,
		environment: environment,
		tags:        make(map[string]string),
		configs:     make(map[string]interface{}),
		versions:    make(map[string]int64),
		callbacks:   make([]func(map[string]interface{}), 0),
	}
}

func (c *ConfigClient) AddTag(key, value string) {
	c.tags[key] = value
}

func (c *ConfigClient) OnChange(callback func(map[string]interface{})) {
	c.callbacks = append(c.callbacks, callback)
}

func (c *ConfigClient) Get(key string) (interface{}, bool) {
	c.mu.RLock()
	defer c.mu.RUnlock()
	val, ok := c.configs[key]
	return val, ok
}

func (c *ConfigClient) GetString(key string) (string, bool) {
	val, ok := c.Get(key)
	if !ok {
		return "", false
	}
	str, ok := val.(string)
	return str, ok
}

func (c *ConfigClient) GetAll() map[string]interface{} {
	c.mu.RLock()
	defer c.mu.RUnlock()
	result := make(map[string]interface{})
	for k, v := range c.configs {
		result[k] = v
	}
	return result
}

func (c *ConfigClient) Start() error {
	if err := c.fetchConfigs(); err != nil {
		return err
	}
	go c.pollLoop()
	return nil
}

func (c *ConfigClient) fetchConfigs() error {
	url := fmt.Sprintf("%s/api/v1/configs?app_id=%s&namespace=%s&environment=%s",
		c.serverURL, c.appID, c.namespace, c.environment)

	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return err
	}

	for k, v := range c.tags {
		req.Header.Set("X-Tag-"+k, v)
	}

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("HTTP %d", resp.StatusCode)
	}

	var configs []struct {
		Key     string      `json:"key"`
		Value   string      `json:"value"`
		Format  string      `json:"format"`
		Version int64       `json:"version"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&configs); err != nil {
		return err
	}

	c.mu.Lock()
	defer c.mu.Unlock()

	changed := false
	for _, cfg := range configs {
		var parsed map[string]interface{}
		switch cfg.Format {
		case "json":
			json.Unmarshal([]byte(cfg.Value), &parsed)
		case "yaml":
			parsed = map[string]interface{}{"raw": cfg.Value}
		default:
			parsed = map[string]interface{}{"raw": cfg.Value}
		}
		c.configs[cfg.Key] = parsed
		c.versions[cfg.Key] = cfg.Version
		changed = true
	}

	if changed && len(c.callbacks) > 0 {
		snapshot := c.GetAll()
		for _, cb := range c.callbacks {
			go cb(snapshot)
		}
	}

	return nil
}

func (c *ConfigClient) pollLoop() {
	for {
		if err := c.longPoll(); err != nil {
			time.Sleep(5 * time.Second)
			continue
		}
	}
}

func (c *ConfigClient) longPoll() error {
	c.mu.RLock()
	versionsCopy := make(map[string]int64)
	for k, v := range c.versions {
		versionsCopy[k] = v
	}
	c.mu.RUnlock()

	reqBody := map[string]interface{}{
		"app_id":       c.appID,
		"namespace":    c.namespace,
		"environment":  c.environment,
		"last_version": versionsCopy,
		"tags":         c.tags,
	}

	data, _ := json.Marshal(reqBody)
	url := fmt.Sprintf("%s/api/v1/long-poll", c.serverURL)

	req, err := http.NewRequest("POST", url, bytes.NewBuffer(data))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")

	ctx, cancel := context.WithTimeout(context.Background(), 35*time.Second)
	defer cancel()
	req = req.WithContext(ctx)

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusNotModified {
		return nil
	}

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("HTTP %d", resp.StatusCode)
	}

	var newConfigs map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&newConfigs); err != nil {
		return err
	}

	c.mu.Lock()
	for k, v := range newConfigs {
		c.configs[k] = v
	}
	c.mu.Unlock()

	snapshot := c.GetAll()
	for _, cb := range c.callbacks {
		go cb(snapshot)
	}

	return nil
}

File: 0508/0508-285/sdk/java/ConfigClient.java
package com.configcenter.sdk;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import okhttp3.*;
import java.io.IOException;
import java.util.*;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.TimeUnit;
import java.util.function.Consumer;

public class ConfigClient {
    private final String serverUrl;
    private final String appId;
    private final String namespace;
    private final String environment;
    private final Map<String, String> tags;
    private final Map<String, Object> configs;
    private final Map<String, Long> versions;
    private final List<Consumer<Map<String, Object>>> callbacks;
    private final OkHttpClient httpClient;
    private final ObjectMapper objectMapper;
    private volatile boolean running;

    public ConfigClient(String serverUrl, String appId, String namespace, String environment) {
        this.serverUrl = serverUrl;
        this.appId = appId;
        this.namespace = namespace;
        this.environment = environment;
        this.tags = new HashMap<>();
        this.configs = new HashMap<>();
        this.versions = new HashMap<>();
        this.callbacks = new CopyOnWriteArrayList<>();
        this.httpClient = new OkHttpClient.Builder()
                .connectTimeout(10, TimeUnit.SECONDS)
                .readTimeout(40, TimeUnit.SECONDS)
                .build();
        this.objectMapper = new ObjectMapper();
    }

    public void addTag(String key, String value) {
        tags.put(key, value);
    }

    public void onChange(Consumer<Map<String, Object>> callback) {
        callbacks.add(callback);
    }

    public Object get(String key) {
        synchronized (configs) {
            return configs.get(key);
        }
    }

    public String getString(String key) {
        Object val = get(key);
        return val != null ? val.toString() : null;
    }

    public Map<String, Object> getAll() {
        synchronized (configs) {
            return new HashMap<>(configs);
        }
    }

    public void start() throws IOException {
        fetchConfigs();
        running = true;
        new Thread(this::pollLoop).start();
    }

    public void stop() {
        running = false;
    }

    private void fetchConfigs() throws IOException {
        String url = String.format("%s/api/v1/configs?app_id=%s&namespace=%s&environment=%s",
                serverUrl, appId, namespace, environment);

        Request.Builder reqBuilder = new Request.Builder().url(url);
        for (Map.Entry<String, String> tag : tags.entrySet()) {
            reqBuilder.header("X-Tag-" + tag.getKey(), tag.getValue());
        }

        try (Response response = httpClient.newCall(reqBuilder.build()).execute()) {
            if (!response.isSuccessful()) {
                throw new IOException("HTTP " + response.code());
            }

            List<ConfigItem> items = objectMapper.readValue(
                    response.body().string(),
                    new TypeReference<List<ConfigItem>>() {}
            );

            boolean changed = false;
            synchronized (configs) {
                for (ConfigItem item : items) {
                    Object parsed = parseValue(item.value, item.format);
                    configs.put(item.key, parsed);
                    versions.put(item.key, item.version);
                    changed = true;
                }
            }

            if (changed && !callbacks.isEmpty()) {
                Map<String, Object> snapshot = getAll();
                for (Consumer<Map<String, Object>> cb : callbacks) {
                    new Thread(() -> cb.accept(snapshot)).start();
                }
            }
        }
    }

    private Object parseValue(String value, String format) {
        try {
            if ("json".equals(format)) {
                return objectMapper.readValue(value, new TypeReference<Map<String, Object>>() {});
            }
        } catch (Exception e) {
            // ignore
        }
        Map<String, Object> raw = new HashMap<>();
        raw.put("raw", value);
        return raw;
    }

    private void pollLoop() {
        while (running) {
            try {
                longPoll();
            } catch (Exception e) {
                try {
                    Thread.sleep(5000);
                } catch (InterruptedException ie) {
                    Thread.currentThread().interrupt();
                    break;
                }
            }
        }
    }

    private void longPoll() throws IOException {
        Map<String, Long> versionsCopy;
        synchronized (configs) {
            versionsCopy = new HashMap<>(versions);
        }

        Map<String, Object> reqBody = new HashMap<>();
        reqBody.put("app_id", appId);
        reqBody.put("namespace", namespace);
        reqBody.put("environment", environment);
        reqBody.put("last_version", versionsCopy);
        reqBody.put("tags", tags);

        String url = serverUrl + "/api/v1/long-poll";
        RequestBody body = RequestBody.create(
                objectMapper.writeValueAsString(reqBody),
                MediaType.parse("application/json")
        );

        Request request = new Request.Builder()
                .url(url)
                .post(body)
                .build();

        try (Response response = httpClient.newCall(request).execute()) {
            if (response.code() == 304) {
                return;
            }
            if (!response.isSuccessful()) {
                throw new IOException("HTTP " + response.code());
            }

            Map<String, Object> newConfigs = objectMapper.readValue(
                    response.body().string(),
                    new TypeReference<Map<String, Object>>() {}
            );

            synchronized (configs) {
                configs.putAll(newConfigs);
            }

            Map<String, Object> snapshot = getAll();
            for (Consumer<Map<String, Object>> cb : callbacks) {
                new Thread(() -> cb.accept(snapshot)).start();
            }
        }
    }

    private static class ConfigItem {
        public String key;
        public String value;
        public String format;
        public long version;
    }
}
